src/EditorFeatures/CSharpTest/Completion/CompletionProviders/Snippets/CSharpForEachSnippetCompletionProviderTests.cs

```diff
@@ -0,0 +1,194 @@
+        public async Task InsertForEachSnippetInGlobalContextTest()
+        {
+            var markupBeforeCommit =
+@"Ins$$
+";
+
+            var expectedCodeAfterCommit =
+@"foreach (var item in collection)
+{$$
+}
+";
+            await VerifyCustomCommitProviderAsync(markupBeforeCommit, ItemToCommit, expectedCodeAfterCommit);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task InserForEachSnippetInConstructorTest()
+        {
+            var markupBeforeCommit =
+@"class Program
+{
+    public Program()
+    {
+        var x = 5;
+        $$
+    }
+}";
+
+            var expectedCodeAfterCommit =
+@"class Program
+{
+    public Program()
+    {
+        var x = 5;
+        foreach (var item in collection)
+        {$$
+        }
+    }
+}";
+            await VerifyCustomCommitProviderAsync(markupBeforeCommit, ItemToCommit, expectedCodeAfterCommit);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task InsertForEachSnippetInLocalFunctionTest()
+        {
+            var markupBeforeCommit =
+@"class Program
+{
+    public void Method()
+    {
+        var x = 5;
+        void LocalMethod()
+        {
+            $$
+        }
+    }
+}";
+
+            var expectedCodeAfterCommit =
+@"class Program
+{
+    public void Method()
+    {
+        var x = 5;
+        void LocalMethod()
+        {
+            foreach (var item in collection)
+            {$$
+            }
+        }
+    }
+}";
+            await VerifyCustomCommitProviderAsync(markupBeforeCommit, ItemToCommit, expectedCodeAfterCommit);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task InsertForEachSnippetInAnonymousFunctionTest()
+        {
+            var markupBeforeCommit =
+@"public delegate void Print(int value);
+static void Main(string[] args)
+{
+    Print print = delegate(int val) {
+        $$
+    };
+}";
+
+            var expectedCodeAfterCommit =
+@"public delegate void Print(int value);
+static void Main(string[] args)
+{
+    Print print = delegate(int val) {
+        foreach (var item in collection)
+        {$$
+        }
+    };
+}";
+            await VerifyCustomCommitProviderAsync(markupBeforeCommit, ItemToCommit, expectedCodeAfterCommit);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task InsertForEachSnippetInParenthesizedLambdaExpressionTest()
+        {
+            var markupBeforeCommit =
+@"Func<int, int, bool> testForEquality = (x, y) =>
+{
+    $$
+    return x == y;
+};";
+
+            var expectedCodeAfterCommit =
+@"Func<int, int, bool> testForEquality = (x, y) =>
+{
+    foreach (var item in collection)
+    {$$
+    }
+
+    return x == y;
+};";
+            await VerifyCustomCommitProviderAsync(markupBeforeCommit, ItemToCommit, expectedCodeAfterCommit);
+        }
+    }
```

Comment:
i was expecting tests where we had a collection in scope, and the foreach then used the name of that collection automatically as the value to foreach over.

Reply:
That's an extension of the current snippets. We want snippets to be even smarter, but for right now I'm trying to match the existing behavior of snippets.

Reply:
https://code.visualstudio.com/docs/editor/userdefinedsnippets#_choice

> Placeholders can have choices as values. The syntax is a comma-separated enumeration of values, enclosed with the pipe-character, for example ${1|one,two,three|}. When the snippet is inserted and the placeholder selected, choices will prompt the user to pick one of the values.

Reply:
I just talked to Taylor about this, this actually doesn't exist in VS - only in VS Code. We would want to file a request to add support for choice.